src/coreclr/tools/aot/ILCompiler/Program.cs

```diff
@@ -713,6 +715,42 @@ static bool IsRelatedToInvalidInput(MethodDesc method)
 
             preinitManager.LogStatistics(logger);
 
+            // If errors were produced (including warnings treated as errors), delete the temporary file
+            // and return error code to avoid misleading build systems into thinking the compilation succeeded.
+            if (logger.HasLoggedErrors)
+            {
+                try
+                {
+                    File.Delete(tempOutputFilePath);
+                }
+                catch
+                {
+                    // If we can't delete the temp file, there's not much we can do.
+                    // The compilation will still fail due to logged errors.
+                }
+
+                return 1;
+            }
+
+            // Rename the temporary file to the final output file
+            try
+            {
+                File.Move(tempOutputFilePath, outputFilePath, overwrite: true);
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine($"Failed to create output file: {ex.Message}");
+                try
+                {
+                    File.Delete(tempOutputFilePath);
+                }
+                catch
+                {
+                    // Ignore cleanup errors
+                }
+                return 1;
+            }
```

Comment:
```suggestion
            // Rename the temporary file to the final output file
            File.Move(tempOutputFilePath, outputFilePath, overwrite: true);
```
We can also leave the exception unhandled. It is out general approach for fatal error handling in ILC. For example, if compilation.Compile fails to open the file or runs out of disk space, it is also going to lead to unhandled exception